cmp/options.go

```diff
@@ -183,9 +183,17 @@ func Ignore() Option {
 // The transformer f must be a function "func(T) R" that converts values of
 // type T to those of type R and is implicitly filtered to input values
 // assignable to T. The transformer must not mutate T in any way.
-// If T and R are the same type, an additional filter must be applied to
-// act as the base case to prevent an infinite recursion applying the same
-// transform to itself (see the SortedSlice example).
+//
+// To help prevent infinite cycles recursively applying the same transform to
+// the output of itself (e.g., in the case where the input and output types
+// are the same), an implicit filter is added such that a transformer is
+// applicable only if that exact transformer is not already in the
+// tail of the Path since the last non-Transform step.
+// While this is sufficient for most usages of Transformers, this does not
```

Comment:
It makes sense to have Transforms and TypeAssertions be part of the Path since they get printed. Otherwise you end up with weird situations when you know that the struct field type was an int, but the Path says its a string because there was a hidden transformation in between. For that reason, there is a one-to-one mapping of recursive calls to cmp.Equal and path steps.

The implementation of cmp.Equal more or less calls itself recursively exactly for each step:
* [`Indirect`](https://github.com/google/go-cmp/blob/d54e85569d47be6e6406055190a66430b6a905b7/cmp/compare.go#L251) for pointers
* [`TypeAssertion`](https://github.com/google/go-cmp/blob/d54e85569d47be6e6406055190a66430b6a905b7/cmp/compare.go#L264) for interfaces
* [`SliceIndex`](https://github.com/google/go-cmp/blob/d54e85569d47be6e6406055190a66430b6a905b7/cmp/compare.go#L470) for slices
* [`MapIndex`](https://github.com/google/go-cmp/blob/d54e85569d47be6e6406055190a66430b6a905b7/cmp/compare.go#L497) for maps
* [`StructField`](https://github.com/google/go-cmp/blob/d54e85569d47be6e6406055190a66430b6a905b7/cmp/compare.go#L540) for structs
* [`Transform`](https://github.com/google/go-cmp/blob/d54e85569d47be6e6406055190a66430b6a905b7/cmp/compare.go#L363) for transformations

The steps `Indirect`, `TypeAssertion`, `SliceIndex`, `MapIndex`, and `StructField` correspond directly with recursive calls dictated by the structure of the value themselves. Only `Transform` is special.